### PR TITLE
BF: fix bug modifying feature dictionary of MROI

### PR DESCRIPTION
**Pull request by matthew-brett**
When iterating through feature dictionary, the hroi module was using the
dictionary keys, which are being modified inside the loop.  Use a static
list of the keys to prevent this happening.

See: https://github.com/nipy/nipy/issues/420

### CODE REVIEW

**Comment by bthirion** [comment]
The change seems innocuous, so this is good for me.

**Comment by matthew-brett** [comment]
Thanks for the review.

**Comment by VirgileFritsch** [comment]
> the hroi module was using the dictionary keys, which are being modified inside the loop

This should have therefore been a problem with every Python version.
I was thinking that the issue was related to the elements of the dictionnary being selected using their position instead of their key.

File: nipy/labs/spatial_models/mroi.py [hunk 1]
```diff
@@ -700,14 +700,14 @@ def select_roi(self, id_list):
         # set new features
         # (it's ok to do that after labels and id modification since we are
         # poping out the former features and use the former id indices)
-        for fid in self.features:
+        for fid in list(self.features):
             f = self.remove_feature(fid)
             sf = [f[id] for id in id_list_pos]
```
Comment (VirgileFritsch) [inline nipy/labs/spatial_models/mroi.py]: `id_list_pos` is a list of "positions" of the dictionnary elements that should be looped over. This code should break if the order of the dictionnary elements cannot be relied upon.

File: nipy/labs/spatial_models/mroi.py [hunk 2]
```diff
@@ -700,14 +700,14 @@ def select_roi(self, id_list):
...
             self.set_feature(fid, sf)
         # set new ROI features
         # (it's ok to do that after labels and id modification since we are
         # poping out the former features and use the former id indices)
-        for fid in self.roi_features:
+        for fid in list(self.roi_features):
             if fid != 'id':
                 f = self.remove_roi_feature(fid)
                 sf = np.ravel(f[int(id_list_pos)])
```
Comment (VirgileFritsch) [inline nipy/labs/spatial_models/mroi.py]: I am being doubtful about `id_list_pos` here again.

**Comment by matthew-brett** [comment]
Virgile - did you read http://bugs.python.org/issue29420? - it records the difference in iteration behavior between the Python versions.   The basic message is - modifying the dictionary while iterating generates undefined behavior - and that undefined behavior changed.

File: nipy/labs/spatial_models/mroi.py (hunk 1 above)
Comment (matthew-brett) [inline nipy/labs/spatial_models/mroi.py]: Virgile - it looks to me as if the output from `f = self.remove_feature(fid)` is a list.  So maybe the `fid` is badly named, it should be called `feature_name` where the feature value is a list of arrays.  Is that right?

**Comment by VirgileFritsch** [comment]
I just did. Still a bit confused by the "difference-in-interation" issue and the "no-order-guarantied" one but it does not matter as far as the fix is ok.

Maybe you want to stay aware that somewhere in the mroi code, we access to the values of a dictionnary by their positions and not their key. It seems to me that this is not a good thing... but I might be wrong :)

Thank you @matthew-brett and @bthirion for taking care of the issue.

**Comment by matthew-brett** [comment]
Can you find a place in the code where we are accessing the items in the dictionary by their position?   I did have a quick look, but found no obvious place.

**Comment by matthew-brett** [comment]
The failure here was caused entirely by the difference in iteration behavior - hence the fix, which is to stop the loop modifying what gets iterated over.

**Comment by VirgileFritsch** [comment]
> Can you find a place in the code where we are accessing the items in the dictionary by their position?

I did in the diff ;)

**Comment by matthew-brett** [comment]
Did you get my comment on your comment?   I think you are wrong about `id_list_pos` referring to positions of the dictionary keys, I think it refers to positions within the lists that are the `feature` values.

File: nipy/labs/spatial_models/mroi.py (hunk 1 above)
Comment (VirgileFritsch) [inline nipy/labs/spatial_models/mroi.py]: Everything is right. Ok for renaming `fid` into `feature_name`, although I think it is more interesting to rename `f` into ... `current_feature`?

**Comment by VirgileFritsch** [comment]
Right, I was mistaken.
I remember a part of the code *did* rely on the dictionnary order, but maybe I or someone fixed it a long time ago. Everything seems fine then.

**Pull request merged**